SuperTokensIOS/Classes/SuperTokensURLProtocol.swift

```diff
@@ -122,10 +123,24 @@ public class SuperTokensURLProtocol: URLProtocol {
                 )
                 
                 if httpResponse.statusCode == SuperTokens.config!.sessionExpiredStatusCode {
+                    /**
+                    * An API may return a 401 error response even with a valid session, causing a session refresh loop in the interceptor.
+                    * To prevent this infinite loop, we break out of the loop after retrying the original request a specified number of times.
+                    * The maximum number of retry attempts is defined by maxRetryAttemptsForSessionRefresh config variable.
+                    */
+                    if self.sessionRefreshAttempts >= SuperTokens.config!.maxRetryAttemptsForSessionRefresh {
+                        let errorMessage = "Error: Received 401 response from \(String(describing: apiRequest.url)). After refreshing the session and retrying the request \(SuperTokens.config!.maxRetryAttemptsForSessionRefresh ) times, we still received 401 responses. Maximum session refresh limit reached. Breaking out of the refresh loop. Please investigate your API. Consider increasing maxRetryAttemptsForSessionRefresh in the config if needed."
+                        print(errorMessage)
+                        self.resolveToUser(data: nil, response: nil, error: SuperTokensError.maxRetryAttemptsReachedForSessionRefresh(message: errorMessage))
+                        return
+                    }
+
                     mutableRequest = self.removeAuthHeaderIfMatchesLocalToken(_mutableRequest: mutableRequest)
                     SuperTokensURLProtocol.onUnauthorisedResponse(preRequestLocalSessionState: preRequestLocalSessionState, callback: {
                         unauthResponse in
                         
+                        self.sessionRefreshAttempts += 1;
```

Comment:
is `self` per request or global?

Reply:
Based on what I found, it is per request. Also, we already had `requestForRetry` member, which is being used as if it was per request.

Reply:
"Based on what I found" -> can you instead prove it to yourself this is per request some how? Some test, or even manual testing will do.

Reply:
I tested it manually and I can confirm that a new instance of `SuperTokensURLProtocol` is created for every request.

<img width="542" alt="image" src="https://github.com/supertokens/supertokens-ios/assets/22813027/f8ab2989-2be9-4477-b3f8-0385f31bc184">